### PR TITLE
Correctly restore markers when undoing/redoing transactions

### PR DESCRIPTION
**Pull request by nathansobo**
This is a continuation of #52. It fixes issues where markers weren't restored correctly when undoing transactions with multiple changes.

Whenever a buffer change occurs that overlaps or invalidates a marker, we record the previous state of the marker on the patch causing that change. When a buffer patch is undone, we restore any markers that have their previous state saved on the patch object. This replaces a previous approach to synthesizing marker patches ahead of time that was breaking down in the presence of multiple changes.

Fixes atom/bracket-matcher#110

### CODE REVIEW

**Comment by nathansobo** [comment]
@kevinsawicki This should be ready to :metal: when you're ready to publish. Core specs pass with it but I haven't run a full CI pass. Happy to address anything that comes up once it's published but @maxbrunsfeld and I are feeling pretty good about it.

**Comment by kevinsawicki** [comment]
Awesome, I'm ready to go, want me to merge and publish this?

**Comment by nathansobo** [comment]
Shoot, one more change I need to make then yes. We no longer need to change the serialization version.

File: src/history.coffee [hunk 1]
```diff
@@ -36,7 +37,9 @@ class History extends Serializable
       if patch instanceof BufferPatch
         @clearRedoStack()
     else
-      @undoStack.push(patch)
```
Comment (nathansobo) [inline src/history.coffee]: We now wrap everything in a transaction so the history always contains the same kind of object. We thought it was required for our first attempt at fixing this, but it ended up being optional. Still, it's a structural improvement so we decided to leave it.

File: src/history.coffee [hunk 2]
```diff
@@ -126,13 +129,14 @@ class History extends Serializable
       false
 
   groupChangesSinceCheckpoint: (checkpoint) ->
-    groupedTransaction = new Transaction
```
Comment (nathansobo) [inline src/history.coffee]: These changes are now possible due to wrapping everything in a transaction.

**Comment by nathansobo** [comment]
@kevinsawicki Okay this is ready. We did a final review and caught one last issue.

**Comment by kevinsawicki** [comment]
:+1:

**Comment by kevinsawicki** [comment]
This broke some specs in Atom so I unpublished it for now:

```
>> bracket matching
>>   matching bracket insertion
>>     when in language specific scope
>>       string interpolation
>>         it does not wrap the selection in the interpolation brackets when the selection is mutli-line
>>           TypeError: Cannot read property 'tokenAtBufferColumn' of undefined
>>             at TokenizedBuffer.module.exports.TokenizedBuffer.tokenForPosition (/private/var/lib/jenkins/workspace/atom/src/tokenized-buffer.coffee:371:25)
>>             at TokenizedBuffer.module.exports.TokenizedBuffer.scopeDescriptorForPosition (/private/var/lib/jenkins/workspace/atom/src/tokenized-buffer.coffee:367:34)
>>             at DisplayBuffer.module.exports.DisplayBuffer.scopeDescriptorForBufferPosition (/private/var/lib/jenkins/workspace/atom/src/display-buffer.coffee:790:22)
>>             at TextEditor.module.exports.TextEditor.scopeDescriptorForBufferPosition (/private/var/lib/jenkins/workspace/atom/src/text-editor.coffee:2574:20)
>>             at Cursor.module.exports.Cursor.getScopeDescriptor (/private/var/lib/jenkins/workspace/atom/src/cursor.coffee:231:13)
>>             at TagFinder.module.exports.TagFinder.isCursorOnTag (/private/var/lib/jenkins/workspace/atom/node_modules/bracket-matcher/lib/tag-finder.coffee:29:50)
>>             at TagFinder.module.exports.TagFinder.findMatchingTags (/private/var/lib/jenkins/workspace/atom/node_modules/bracket-matcher/lib/tag-finder.coffee:96:29)
>>             at BracketMatcherView.module.exports.BracketMatcherView.updateMatch (/private/var/lib/jenkins/workspace/atom/node_modules/bracket-matcher/lib/bracket-matcher-view.coffee:90:28)
>>             at /private/var/lib/jenkins/workspace/atom/node_modules/bracket-matcher/lib/bracket-matcher-view.coffee:59:8
>>             at Emitter.module.exports.Emitter.emit (/private/var/lib/jenkins/workspace/atom/node_modules/event-kit/lib/emitter.js:82:11)
>>             at /private/var/lib/jenkins/workspace/atom/src/cursor.coffee:48:16
>>             at Emitter.module.exports.Emitter.emit (/private/var/lib/jenkins/workspace/atom/node_modules/event-kit/lib/emitter.js:82:11)
>>             at Marker.module.exports.Marker.notifyObservers (/private/var/lib/jenkins/workspace/atom/src/marker.coffee:381:16)
>>             at /private/var/lib/jenkins/workspace/atom/src/marker.coffee:73:54
>>             at Emitter.module.exports.Emitter.emit (/private/var/lib/jenkins/workspace/atom/node_modules/event-kit/lib/emitter.js:82:11)
>>             at Marker.module.exports.Marker.update (/private/var/lib/jenkins/workspace/atom/node_modules/text-buffer/lib/marker.js:567:22)
>>             at MarkerPatch.module.exports.MarkerPatch.applyTo (/private/var/lib/jenkins/workspace/atom/node_modules/text-buffer/lib/marker-patch.js:50:75)
>>             at Transaction.module.exports.Transaction.applyTo (/private/var/lib/jenkins/workspace/atom/node_modules/text-buffer/lib/transaction.js:60:29)
>>             at History.module.exports.History.undo (/private/var/lib/jenkins/workspace/atom/node_modules/text-buffer/lib/history.js:91:24)
>>             at TextBuffer.module.exports.TextBuffer.undo (/private/var/lib/jenkins/workspace/atom/node_modules/text-buffer/lib/text-buffer.js:679:27)
>>             at TextEditor.module.exports.TextEditor.undo (/private/var/lib/jenkins/workspace/atom/src/text-editor.coffee:1136:13)
>>             at [object Object].<anonymous> (/private/var/lib/jenkins/workspace/atom/node_modules/bracket-matcher/spec/bracket-matcher-spec.coffee:899:18)
>>
>> TextEditor
>>   cursor
>>     autoscroll
>>       it autoscrolls to the cursor when it moves due to undo
>>         Expected 75 to be 0.
>>           at [object Object].<anonymous> (/private/var/lib/jenkins/workspace/atom/spec/text-editor-spec.coffee:995:39)
>>   buffer manipulation
>>     .undo() and .redo()
>>       it restores the selected ranges after undo and redo
>>         Expected [ { start : { row : 1, column : 6 }, end : { row : 1, column : 6 } }, { start : { row : 1, column : 19 }, end : { row : 1, column : 19 } } ] to equal [ [ [ 1, 6 ], [ 1, 6 ] ], [ [ 1, 18 ], [ 1, 18 ] ] ].
>>           at [object Object].<anonymous> (/private/var/lib/jenkins/workspace/atom/spec/text-editor-spec.coffee:3182:50)
>>         Expected [ { start : { row : 1, column : 6 }, end : { row : 1, column : 10 } }, { start : { row : 1, column : 26 }, end : { row : 1, column : 31 } } ] to equal [ [ [ 1, 6 ], [ 1, 10 ] ], [ [ 1, 22 ], [ 1, 27 ] ] ].
>>           at [object Object].<anonymous> (/private/var/lib/jenkins/workspace/atom/spec/text-editor-spec.coffee:3185:50)
```

**Pull request merged**